app/rdf/light-cube.ts

```diff
@@ -0,0 +1,266 @@
+import groupBy from "lodash/groupBy";
+import ParsingClient from "sparql-http-client/ParsingClient";
+
+import {
+  BaseComponent,
+  BaseDimension,
+  DataCubeMetadata,
+  Dimension,
+  Measure,
+  TemporalDimension,
+} from "@/domain/data";
+import { resolveDimensionType, resolveMeasureType } from "@/graphql/resolvers";
+import * as ns from "@/rdf/namespace";
+import {
+  getDataKind,
+  getIsNumerical,
+  getScaleType,
+  getTimeFormat,
+  getTimeUnit,
+  parseNumericalTerm,
+  parseResolution,
+} from "@/rdf/parse";
+import { getCubeMetadata } from "@/rdf/query-cube-metadata";
+import { buildLocalizedSubQuery } from "@/rdf/query-utils";
+
+type LightCubeOptions = {
+  iri: string;
+  locale: string;
+  sparqlClient: ParsingClient;
+};
+
+export class LightCube {
+  public iri: string;
+  private locale: string;
+  public metadata: DataCubeMetadata | undefined;
+  public dimensions: Dimension[] = [];
+  public measures: Measure[] = [];
+  private sparqlClient: ParsingClient;
+
+  constructor(options: LightCubeOptions) {
+    const { iri, locale, sparqlClient } = options;
+    this.iri = iri;
+    this.locale = locale;
+    this.sparqlClient = sparqlClient;
+  }
+
+  /** Use to promote the cube to newest version. */
+  public async promote(latest: boolean) {
+    if (!latest) {
+      return this;
+    }
+
+    const query = `PREFIX schema: <http://schema.org/>
+
+SELECT ?iri WHERE {
+  ?versionHistory schema:hasPart  <${this.iri}> .
+  ?versionHistory schema:hasPart ?iri .
+  ?iri schema:version ?version .
+}
+ORDER BY DESC(?version)
+LIMIT 1`;
```

Comment:
IIUC this would upgrade to any latest version - even if the later version is only a draft while the current version is published. This arises the question, how can we ensure to upgrade from published to published // or from draft|published to draft|published. (or maybe you ensure this already by other means)

... it might be useful to incorporate such a query (together with a second clause for cubes w/o versionHistory) within the main query itself i.e. with a nested SELECT pattern

you may try to nest i.e. https://s.zazuko.com/3kWrvgW
resulting in i.e. 

```sparql
PREFIX schema: <http://schema.org/>

SELECT ?cube ?value ?etc WHERE {

# main query 

{ #start nested upgrade query
SELECT ?iri WHERE {
  {
    VALUES ?iri { <https://environment.ld.admin.ch/foen/nfi/nfi_C-22/cube/2023-1> }
    ?iri schema:version ?version .
  }
  UNION
  {
    VALUES ?oldIri { <https://environment.ld.admin.ch/foen/nfi/nfi_C-22/cube/2023-1> }
    ?versionHistory schema:hasPart ?oldIri .
    ?versionHistory schema:hasPart ?iri .
    ?iri schema:version ?version .
    ?iri schema:creativeWorkStatus ?status .
    ?oldIri schema:creativeWorkStatus ?oldStatus .
    FILTER(?status IN ( <https://ld.admin.ch/vocabulary/CreativeWorkStatus/Published>, ?oldStatus))
  }
}
ORDER BY DESC(?version)
LIMIT 1
} #end nested upgrade query

# rest of the main query ...

}
```

Reply:
Thanks for pointing this out @Rdataflow and proving an example query! I am not sure if we should have a fallback for cubes without a version history, as we only considered ones with it in our current, very old logic.

You can see how we were doing this until now: https://github.com/visualize-admin/visualization-tool/blob/90c7a75bb7ee0fd4c27220c162067626bfdc8d79/app/rdf/queries.ts#L54-L114 (I am going to replicate this logic in a subquery in `LightCube`)

Reply:
Should be done via https://github.com/visualize-admin/visualization-tool/pull/1378/commits/12bef541c66a3297f44da2542ab5d345d2f0c4ab 🤞

Reply:
@bprusinowski the good news is: the provided subquery already works for cubes w/o versionHistory thanks to the first set just before UNION. (and there are plenty of them, i.e. the electricity-price cubes. they work well on visualize TEST currently and will continue to work with the provided pattern while the others with history upgrade well on TEST and the new pattern as well...)

how to test - case w/o versionHistory
1. open [electricity price](https://test.visualize.admin.ch/browse?dataset=https%3A%2F%2Fenergy.ld.admin.ch%2Felcom%2Felectricityprice&dataSource=Prod)
2. observe SPARQL queries
3. observe visualize works perfectly

how to test - case upgrade with versionHistory
1. open expired cube [nfi cube 2023.1](https://visualize.admin.ch/create/new?cube=https://environment.ld.admin.ch/foen/nfi/nfi_C-90/cube/2023-1) in visualize TEST or PROD
2. observe sparql queries
3. observe visualize seamlessly upgrades to version 2023.3

I hope this helps you to understand
- the need for the ongoing support for cubes without versionHistory.
- a working solution with a simple subquery pattern handling upgrades where applicable - while reducing the amounts of queries.

Just ping me, in case you encounter any questions while implementing...

Reply:
@Rdataflow I think the current implementation should work in the same manner, as if the query does not find an iri, it keeps the current one which we already have when initializing the cube (https://github.com/visualize-admin/visualization-tool/blob/b6f56ba91e86f6e75e0e7d2d0ef1e230b410ae90/app/rdf/light-cube.ts#L60-L68) – it means that we don't have to do a round trip and try to fetch cube without version history, because we've already done so in a previous step.

Is there some other use case you see where the current approach won't work?

Reply:
See that the cube without version history works in the preview deployment: https://visualization-tool-git-perf-optimize-data-cube-comp-4a3cd4-ixt1.vercel.app/en/create/new?cube=https://energy.ld.admin.ch/elcom/electricityprice-canton&dataSource=Prod